Adding tutorial for confidence ensembles

Kipok opened the pull request:
# What does this PR do?

There are a couple of bugfixes to the confidence computation and also a tutorial.

It currently misses a few links:
- to the paper (already submitted to arxiv, waiting to get the link)
- to the confidence tutorial from Aleksandr, which is not merged yet
- to the documentation of the confidence ensembles, which I hope to update in the next few days either in this PR or in the follow-up

**Collection**: ASR

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.


## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

Kipok commented:
Update - fixed all links, except for the confidence tutorial, added docs and also a few simple standalone tests. Bug fixes are removed, since they are merged in another PR

GNroy reviewed (commented):
This is a preliminary reviewas I was unable to finish the Colab run.
Overall, great PR!
Please confider the following comments.
It would also be great to somehow highlight the difference between the last and penultimate cells, as they long but mostly similar.

Kipok commented:
Thanks for the review, @GNroy! I addressed all of the feedback - also changed the history to remove the image I originally pushed and moved it over to the release files, so that it's accessible in colab and not in the git history.

Sorry about the colab issue - only tested locally because was developing that way. I now verified that it works well in both colab and local mode. Just make sure to restart the kernel after running the first cell to pick up the installed nemo package (there is a comment at the bottom of that cell explaining this)

GNroy reviewed (approved):
LGTM, all my comments were addressed.
Thanks!

titu1994 commented on tests/collections/asr/test_asr_interctc_models.py:
```diff
@@ -86,7 +86,6 @@ class TestInterCTCLoss:
             ([], [0.3]),
         ],
     )
-    @pytest.mark.pleasefixme
```
Was the threshold or stuff fixed? This caused test failures before

Kipok replied:
It was fixed already, just for some reason there was this decorator applied 2 times and I only removed one of them in the original PR

titu1994 replied:
Lol ok

titu1994 commented on tutorials/asr/Confidence_Ensembles.ipynb:
Seems a few images in the tutorial are missing. You can add them to the latest release

titu1994 replied:
Could you put the SDP download and install at the very top of the notebook

Kipok replied:
Which images are missing?

titu1994 replied:
Oh in the preview the image in the first section of the tutorial (explanation of confidence) was missing. Maybe it will be fixed upob merge c

titu1994 reviewed (dismissed):
The docs and code looks great, minor comments about the tutorial but overall I only skimmed it so please do get a thorough review of it

titu1994 commented on tutorials/asr/Confidence_Ensembles.ipynb:
```diff
@@ -0,0 +1,517 @@
+    "\n",
+    "A short answer — you can use any ASR models. E.g., you can combine a number of CTC models, or Transducer models, or even mix-and-match. \n",
+    "\n",
+    "A more detailed answer is that hte performance of the confidence ensemble is upper-bounded by the performance of the best model on each of the input examples. Thus you will benefit if some of your models work really well on part of the input compared to other models. This way you will get more gains compared to each separate model, and it will also make correct model identification easier.\n ...
```
Imo this section is quite heavy to read in one sitting. Breaking apart the subsection into digestible chunks may be better c

This is fine as well, just a bit heavy.

Kipok replied:
I was not sure how much details to put, so decided to put enough for people to understand a bit more than just examples, but also encourage to skip around the text and go straight to code and return if they want to learn more. Maybe I should make it more explicit - right now I just say 

> Each cell is mostly self-contained, so feel free to skip around or jump directly to the code part if you want to see usage examples right away.

at the end of the first text cell.

titu1994 commented on tutorials/asr/Confidence_Ensembles.ipynb:
```diff
@@ -0,0 +1,517 @@
+    "Let's now talk about the \"model selection block\". First of all — you don't need to know the details to use confidence ensembles, calibration is always automatically performed when you build the model. But if you want to learn more, read on!\n",
+    "\n",
+    "First, let's discuss why we need a separate \"model selection block\" to pick the most confident model. If we had an access to the perfect confidence, which would exactly equal to the probability of the model's output being correct, we wouldn't need this block. In this idealized case we can simply take the model with the maximum confidence score. But in practice, models tend to be over- or u ...
+    "\n",
```
Wall of text, hard to maintain concentration for so much text. Maybe split apart or use some form of visual aid to give a high level understanding then dig into details. 

It's fine for now but very heavy section.

Kipok replied:
Yeah, it's a bit too much, but that's why I have

> First of all — you don't need to know the details to use confidence ensembles, calibration is always automatically performed when you build the model. But if you want to learn more, read on!

in the first paragraph :)

titu1994 reviewed (approved):
The tutorial is excellently written and documented. Great work ! 

Maybe it can be improved with more pictures or some other form of explanation in the text heavy sections but it's also ok as is.
